{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "232003af",
   "metadata": {},
   "source": [
    "# Generating Multiple Sequence Alignement given PDB structure\n",
    "\n",
    "Given a PDB structure file we use ProDy to search all available Pfam alignments for the best matching Multiple Sequence Alignment (MSA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a8f1c5bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initial Import\n",
    "\n",
    "# System packages\n",
    "import os.path, sys\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "# Scientific Computing packages\n",
    "import numpy as np\n",
    "np.random.seed(1) # set random seed for reproducibility\n",
    "import pandas as pd\n",
    "from scipy import linalg\n",
    "from scipy.sparse import csr_matrix\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from scipy.spatial import distance_matrix, distance\n",
    "\n",
    "\n",
    "# Biopython packages\n",
    "import Bio.PDB, warnings\n",
    "pdb_list = Bio.PDB.PDBList()\n",
    "pdb_parser = Bio.PDB.PDBParser()\n",
    "from Bio import BiopythonWarning\n",
    "warnings.simplefilter('ignore', BiopythonWarning)\n",
    "from Bio import SeqIO, pairwise2\n",
    "\n",
    "# Parallelization and computing diagnosti packages\n",
    "from joblib import Parallel, delayed\n",
    "import timeit\n",
    "\n",
    "# Plotting packages\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# # --- Import our Code ---# #\n",
    "# Direct Coupling Analysis code\n",
    "from direct_info import direct_info\n",
    "\n",
    "# Import data processing and general DCA_ER tools\n",
    "from data_processing import data_processing, pdb2msa\n",
    "import tools\n",
    "# # -----------------------# #\n",
    "\n",
    "# Import ProDy code.\n",
    "from prody import *\n",
    "\n",
    "# Define Computational Resources\n",
    "n_cpus = 10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebac6446",
   "metadata": {},
   "source": [
    "## Define Data directories\n",
    "* data_path: Where is the MSA and pdb structure data\n",
    "* out_dir: Where do you want the processed MSA data to go\n",
    "* pdb_dir: Where do you want the PDB-MSA reference and structure data to go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1eb41316",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Define data directories -- NEED TO DEFINE\n",
    "data_path = Path('example_protein_data')\n",
    "DCA_ER_dir = '.' # Set DCA_ER directory\n",
    "out_dir = '%s/protein_data/di/' % DCA_ER_dir\n",
    "processed_data_dir = \"%s/protein_data/data_processing_output\" % DCA_ER_dir\n",
    "pdb_dir = '%s/protein_data/pdb_data/' % DCA_ER_dir\n",
    "\n",
    "\n",
    "if not os.path.exists(out_dir):\n",
    "    os.makedirs(out_dir)\n",
    "    os.makedirs(processed_data_dir)\n",
    "    os.makedirs(pdb_dir)\n",
    "\n",
    "    \n",
    "# Path to gzipped pdb file\n",
    "pdb_path = \"/pdb/pdb/zd/pdb1zdr.ent.gz\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce00c518",
   "metadata": {},
   "source": [
    "## Un-zip and load PDB file to get PDB structure (1zdr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "28fc4ee8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unzipping /pdb/pdb/zd/pdb1zdr.ent.gz to ./protein_data/pdb_data/pdb1zdr.ent\n"
     ]
    }
   ],
   "source": [
    "import gzip, shutil\n",
    "def gunzip(file_path, output_path):\n",
    "    print('Unzipping %s to %s' % (file_path, output_path))\n",
    "    with gzip.open(file_path,\"rb\") as f_in, open(output_path,\"wb\") as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)\n",
    "\n",
    "        \n",
    "        \n",
    "unzipped_pdb_filename = os.path.basename(pdb_path).replace(\".gz\", \"\")\n",
    "pdb_out_path = \"%s%s\" % (pdb_dir, unzipped_pdb_filename)\n",
    "gunzip(pdb_path, pdb_out_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3b19927",
   "metadata": {},
   "source": [
    "## Generate PDB2MSA Data\n",
    "* For each available polypeptide sequence in the PDB structure us ProDy to BLAST for corresponding sequences in the Pfam domain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3799b601",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDB2MSA Results:\n",
      " [   Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1  Unnamed: 0.1.1.1  \\\n",
      "0           0             0               0                 0   \n",
      "1           1             1               1                 1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1  Unnamed: 0.1.1.1.1.1  Unnamed: 0.1.1.1.1.1.1  \\\n",
      "0                   0                     0                       0   \n",
      "1                   1                     1                       1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1.1.1.1  Unnamed: 0.1.1.1.1.1.1.1.1  \\\n",
      "0                         0                           0   \n",
      "1                         1                           1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1.1.1.1.1.1  ...  ali_end ali_start bitscore  end  \\\n",
      "0                             0  ...      160         1   215.11  160   \n",
      "1                             1  ...      160         1   215.08  160   \n",
      "\n",
      "    cond_evalue    ind_evalue    evidence hmm_end hmm_start start  \n",
      "0  2.300000e-68  4.500000e-64  hmmer v3.0     160         1     1  \n",
      "1  2.300000e-68  4.600000e-64  hmmer v3.0     160         1     1  \n",
      "\n",
      "[2 rows x 30 columns]]\n",
      "\n",
      "\n",
      "\n",
      "PDB DF with associated Protein Families\n",
      "    Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1  Unnamed: 0.1.1.1  \\\n",
      "0           0             0               0                 0   \n",
      "1           1             1               1                 1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1  Unnamed: 0.1.1.1.1.1  Unnamed: 0.1.1.1.1.1.1  \\\n",
      "0                   0                     0                       0   \n",
      "1                   1                     1                       1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1.1.1.1  Unnamed: 0.1.1.1.1.1.1.1.1  \\\n",
      "0                         0                           0   \n",
      "1                         1                           1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1.1.1.1.1.1  ...  ali_end ali_start bitscore  end  \\\n",
      "0                             0  ...      160         1   215.11  160   \n",
      "1                             1  ...      160         1   215.08  160   \n",
      "\n",
      "    cond_evalue    ind_evalue    evidence hmm_end hmm_start  start  \n",
      "0  2.300000e-68  4.500000e-64  hmmer v3.0     160         1      1  \n",
      "1  2.300000e-68  4.600000e-64  hmmer v3.0     160         1      1  \n",
      "\n",
      "[2 rows x 29 columns]\n"
     ]
    }
   ],
   "source": [
    "pdb2msa_results = pdb2msa(pdb_out_path, pdb_dir, create_new=False)\n",
    "print('PDB2MSA Results:\\n', pdb2msa_results)\n",
    "\n",
    "if len(pdb2msa_results) > 1:\n",
    "    fasta_file = pdb2msa_results[0]\n",
    "    prody_df = pdb2msa_results[1]\n",
    "else:\n",
    "    prody_df = pdb2msa_results[0]\n",
    "\n",
    "\n",
    "print('\\n\\n\\nPDB DF with associated Protein Families\\n', prody_df.loc[:,  [column for column in prody_df.columns if column not in ['locations', 'PDB Sequence']]].head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef76b443",
   "metadata": {},
   "source": [
    "## Load and Pre-Process Pfam-MSA\n",
    "* With the PDB2MSA data we can now loop through all the different PDB-MSA pairings for all available sequences\n",
    "* Because the PDB2MSA data is ordered by PDB-polypeptide sequence to MSA alignment score the we will loop from most similar to least similar as we load and Pre-Process the different MSAs for our analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4c45f4c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "Looping through Prody Search DataFrame:\n",
      "    Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1  Unnamed: 0.1.1.1  \\\n",
      "0           0             0               0                 0   \n",
      "1           1             1               1                 1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1  Unnamed: 0.1.1.1.1.1  Unnamed: 0.1.1.1.1.1.1  \\\n",
      "0                   0                     0                       0   \n",
      "1                   1                     1                       1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1.1.1.1  Unnamed: 0.1.1.1.1.1.1.1.1  \\\n",
      "0                         0                           0   \n",
      "1                         1                           1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1.1.1.1.1.1  ...  ali_end ali_start bitscore  end  \\\n",
      "0                             0  ...      160         1   215.11  160   \n",
      "1                             1  ...      160         1   215.08  160   \n",
      "\n",
      "    cond_evalue    ind_evalue    evidence hmm_end hmm_start start  \n",
      "0  2.300000e-68  4.500000e-64  hmmer v3.0     160         1     1  \n",
      "1  2.300000e-68  4.600000e-64  hmmer v3.0     160         1     1  \n",
      "\n",
      "[2 rows x 30 columns]\n",
      "\n",
      "\n",
      "Getting msa with following pdb2msa entry:\n",
      " (0, Unnamed: 0                                                                        0\n",
      "Unnamed: 0.1                                                                      0\n",
      "Unnamed: 0.1.1                                                                    0\n",
      "Unnamed: 0.1.1.1                                                                  0\n",
      "Unnamed: 0.1.1.1.1                                                                0\n",
      "Unnamed: 0.1.1.1.1.1                                                              0\n",
      "Unnamed: 0.1.1.1.1.1.1                                                            0\n",
      "Unnamed: 0.1.1.1.1.1.1.1                                                          0\n",
      "Unnamed: 0.1.1.1.1.1.1.1.1                                                        0\n",
      "Unnamed: 0.1.1.1.1.1.1.1.1.1                                                      0\n",
      "Unnamed: 0.1.1.1.1.1.1.1.1.1.1                                                    0\n",
      "PDB ID                                                                         1zdr\n",
      "Chain                                                                             B\n",
      "Polypeptide Index                                                                 0\n",
      "Pfam                                                                        PF00186\n",
      "accession                                                                PF00186.22\n",
      "class                                                                        Domain\n",
      "id                                                                           DHFR_1\n",
      "type                                                                         Pfam-A\n",
      "PDB Sequence                      MISHIVAMDENRVIGKDNRLPWHLPADLAYFKRVTMGHAIVMGRKT...\n",
      "ali_end                                                                         160\n",
      "ali_start                                                                         1\n",
      "bitscore                                                                     215.11\n",
      "end                                                                             160\n",
      "cond_evalue                                                                     0.0\n",
      "ind_evalue                                                                      0.0\n",
      "evidence                                                                 hmmer v3.0\n",
      "hmm_end                                                                         160\n",
      "hmm_start                                                                         1\n",
      "start                                                                             1\n",
      "Name: 0, dtype: object)\n",
      "PDB ID: 1zdr, Pfam ID: PF00186\n",
      "Original Sequence length:  918\n",
      "\n",
      "\n",
      "#--------------------- Find PDB Sequence in MSA ---------------#\n",
      "Lets find the reference sequence in this MSA\n",
      "...We want index of the MSA sequence which best matches our PDB-polypeptide sequence...?\n",
      "PDB-polypeptids sequence (len 160):\n",
      "MISHIVAMDENRVIGKDNRLPWHLPADLAYFKRVTMGHAIVMGRKTFEAIGRPLPGRDNVVVTGNRSFRPEGCLVLHSLEEVKQWIASRADEVFIIGGAELFRATMPIVDRLYVTKIFASFPGDTFYPPISDDEWEIVSYTPGGKDEKNPYEHAFIIYER\n",
      "\n",
      "\n",
      "Looping through MSA sequences to find best matching sequence for our MSA reference sequence!\n",
      "    ... found a match upgrade at index  0\n",
      "MI---SH-IV---AMD-ENRVIGKD-NRLPWHLPADLAYFKRVTMGHAIVM-GRK-T---FEAIG----RPLPG-RD-NVVV-TGNR---SFR-PEGC--LVLH-S----L---E------EVKQWIASRADEVFIIGGAE--LFR---ATM------PIVDRLYV-TK-IFAS---FPGD--TFYP-PISD-DEWE----I----V-SYTPGGKDEKNP-Y----EHAFII----YER-\n",
      "|    |  |    |   |    ||  |          |    |     |   ||  |   |        ||||  |  |||| |      |   |||   ||   |    |   |      ||        ||||||||    |     |        | | |||  |  |      |  |  ||   | |  | ||    |    | |      |   | |    |  |      ||  \n",
      "M-AFFS-RI-TSTA--QE----GK-KN----------A----V-----I--LGR-RTWLSF----PPKFRPLP-NR-VNVVVST---QLES--VPEG-AYLV--KSFEDSLRLVESLVDGGEV--------DEVFIIGG--HGL--YKEA--LEQEVYP-V-RLY-YT-HI---MKDF--DCDTF--FP-S-VD-WERFTPIQLATVDS------D---PKYCGDVE--F--RFAVYE-K\n",
      "  Score=72\n",
      "\n",
      "    ... found a match upgrade at index  1\n",
      "MISHI-V-AMD-EN-RV--IGKD--NR-LPWHLPA-DLAY-F-KRV-T-MGHAIV-M--GRKTFEAI--G-RPL-PGRDNVV---V-T-----G-NRSFRP-EGCL-VL-HS-LEEVKQWIASR----A-D-E--VFIIGGAE----LFRATM-PIV--D---RLYVTK-IFAS---FPG-DT-FYP-P-ISDDEW-EI---VSYTPGGK------DEKN---PYEHAFII-YER---\n",
      "|||   | |   |  |   ||    |  ||||||  ||   | |   | |||    |  ||||||    | | | |||       | |     | |      || | |  |  |||            | | |  |  ||||     ||     |    |   |   |  |      ||  |  |   | |  ||  |    |      |      || |   |  |     ||    \n",
      "MIS--FVYA--QE-SR-GGIG--YQN-QLPWHLP-NDL--KFFK--ETTMGH---TMLMGRKTFE--SMGQR-LLPGR----KTIVMTRDLNYGKN-----IEG-LTV-IH-RLEE--------GLELAKDQELMV--IGGA-DIFKLF----LP--YADQIIR---T-LI---EREFP-SD-VF--MPAI--DE-TE-WQLV------KVVAGILDE-NNTIP--H----QYE-WWE\n",
      "  Score=82\n",
      "\n",
      "    ... found a match upgrade at index  4\n",
      "MISH----IVAMDENR----VIGK-DNR-LPWH-LPA-DLAYF-KRVT-MGHA-IV-MGRKTF-EA-IGRPLPGRD-NV-VVTGNR-SF--RP-EGCL--VLH-SLEEVKQWI--ASR-A----D---EV-FI-IGGAE-LF------RAT--MPIVDRLYVTK-IF-AS-FPGD--T-FYP--PISDDE---W-EIVSYTPGGKD-EKNPYEH-A-----FII------YER\n",
      "        |||    |    |||  ||  |||  ||  ||| | || | || | |  |||||  |  ||||||||  |  |||  | |   |  |||   |   |       |  |   |    |   |  |  ||||  |       ||   | |     ||  |  |  |  |  | |    |   |    | | ||        |  |  | |     |        || \n",
      "----TLTLIVA----RARNGVIG-RDN-QLPW-RLP-EDLA-FFKR-TTMG-API-IMGRKT-HE-SIGRPLPGR-RN-IVVT--RDS-TLR-FEGC-DTV--TS-------IDDA--LALAAADGAAE-AF-VIGGA-QL-YGDAIGRA-RKM-I-----VT-EI-DA-DF--DGNTRF--EAP---D-ASQWRE-VS-------RE--P--HRAAPPNDF--GFAFVTYE-\n",
      "  Score=87\n",
      "\n",
      "    ... found a match upgrade at index  7\n",
      "MISH----IVAMDENR----VIGK-DNR-LPWH-LPA-DLAYF-KRVT-MGHA-IV-MGRKTF-EA-IGRPLPGRD-NV-VVTGN-R--SFRPE---GCL-VLH--SLEEVKQWI-ASRA------DEV-FI-IGGAE-L----F-RATMPIVD-RLY-VTK-IF-AS-FP-GDT-FYP-PISDDE---W-EIVSY--T----PGGKDEKNPYEH-AFII------YER\n",
      "        |||    |    |||  ||  |||  ||  ||| | || | || | |  |||||  |  ||||||||  |  ||    |  | |     ||  |    ||        |  |      ||  |  ||||  |    | ||       ||  ||  |  |  |  ||| | | |   |    | | ||   |    |       |    ||        |||\n",
      "----TLTLIVA----RARNGVIG-RDN-QLPW-RLP-EDLA-FFKR-TTMG-API-IMGRKT-HE-SIGRPLPGR-RN-IVV---SRDAS-R--RYDGC-DV--VTSL-------DA--ALALAPADE-AF-LIGGA-QLYRDGFERA------NRL-IVT-EI-DA-DF-DGDTRF-PAP---D-PAHWRE-VS-RETHHAAP-------P---NAF--DFAFVTYER\n",
      "  Score=89\n",
      "\n",
      "    ... found a match upgrade at index  25\n",
      "MI-SH-IVAMDEN-RV-IGKD-NR--LPWH-LPA-DL-AYF-KRVTM--GHAI-V-MGRKTFEA-IGR-PLPG-RD-NVVV-TGNR--SF-RP-EGCL--VLH-SLEEVKQWI--AS--RA------D-EV-FI-IGGAE---LFRATMPIVDR----LYVTK-IFAS--FPG-DTFYP-P-ISDDE---WEI-VSYTPG----G-KD-EK-NPY--EHAFI-IYER\n",
      " | |  ||||||  |  ||   |   |||  ||  |  | | || |   ||   | |||||||  ||  |||  |  | || |  |  |  |  |||   |   |   |      |   |       | |  |  |||||   ||   ||        ||||  | |   | | |||   | |  |    |   |  |      | |  |  |||  |  |  ||  \n",
      "-IIS-MIVAMDE-QR-GIG--IN-NKLPW-RLP-ED-TA-FFKR-T-TIGH--TVLMGRKTFE-SIG-KPLP-KR-TN-VVLT--RDRS-YR-AEGC-EMV--YS---V----DEA-LER-YRNGVPDDE-LF-VIGGAEVYKLF---MP----HAAKLYVT-EI-A-HRF-GADTF--FPEI--D-SSLW--KV--T--NRQQGVK-NE-SNPYDYE--F-VIY--\n",
      "  Score=93\n",
      "\n",
      "    ... found a match upgrade at index  27\n",
      "M-ISH-IV-AMDE-NRV-IGKD-NR--LPWHLPA-DLAY-F-KR-VTMGHA-IV-MGRKTF-EA-IGR-PLPGRD-NVVV--TG-NRS--FRP-EGCL-VLHSL--E--EVKQWIAS-RAD-EV-FI-IGGAE---LFRA-TMPIVDRL-YV-TK-I---F-ASFPGDTFYPPI-SD-DE---WEI--VSYTPG---GKDE---KNPYEHA--FII--YER\n",
      "  ||  |  |||  ||  ||   |   ||||||  ||   | |  |||||  |  |||||  |  ||  |||||  |  |  |  |    ||  |||  ||||   |  |     |  | | |  |  |||||   || | |   |||  |  |  |   | |    || |    || |    |    ||       |  |   |||||    |    |||\n",
      "-SIS-FI-FAMD-RNR-AIG--VN-NQLPWHLP-GDL--KFFK-SVTMGH-PI-LMGRKT-YE-SIG-KPLPGR-RN--VILT-QN--TAFR-AEGC-EVLHS-VGEALE-----A-FR-DQE-LF-VIGGAEVFKLF-AST---VDR-MY-IT-FIEHEFEA----DT-Y---MSDLD-LSAW--TLVS----SEQG--ERNEKNPYE--YYF--RVYER\n",
      "  Score=97\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    ... found a match upgrade at index  69\n",
      "MISHIVAMDENRVIGKDNRLPWHLPADLAYFKRVTMGHAIVMGRKTFEAIGRPLPGRDNVVVTG-NRSFRPEGCLVLHSLEEVKQWIASRADEVFIIGGAELFRATMPIVDRLYVTKIFASFPGDTFYPPISDDEWEIVSYTPGGKDEKNPYEHAFIIYER\n",
      "|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||  ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "MISHIVAMDENRVIGKDNRLPWHLPADLAYFKRVTMGHAIVMGRKTFEAIGRPLPGRDNVVVT-RNRSFRPEGCLVLHSLEEVKQWIASRADEVFIIGGAELFRATMPIVDRLYVTKIFASFPGDTFYPPISDDEWEIVSYTPGGKDEKNPYEHAFIIYER\n",
      "  Score=159\n",
      "\n",
      "Best match is sequence 69\n",
      "    Hamming distance of 160 (length 160)\n",
      "\n",
      "\n",
      "1 [63]\n",
      "\n",
      "Aligned PDB and ref seq:\n",
      "PDB polypeptide sequence (length 159):  MISHIVAMDENRVIGKDNRLPWHLPADLAYFKRVTMGHAIVMGRKTFEAIGRPLPGRDNVVVTNRSFRPEGCLVLHSLEEVKQWIASRADEVFIIGGAELFRATMPIVDRLYVTKIFASFPGDTFYPPISDDEWEIVSYTPGGKDEKNPYEHAFIIYER\n",
      "MSA-matched sequence     (length 159):  MISHIVAMDENRVIGKDNRLPWHLPADLAYFKRVTMGHAIVMGRKTFEAIGRPLPGRDNVVVTNRSFRPEGCLVLHSLEEVKQWIASRADEVFIIGGAELFRATMPIVDRLYVTKIFASFPGDTFYPPISDDEWEIVSYTPGGKDEKNPYEHAFIIYER\n",
      "\n",
      "\n",
      "Trimming MSA by reference sequence (s -> s_tpr)...\n",
      "dimensions of s_tbr: (7750, 160)\n",
      "s_tbr[tpdb=69] =  MISHIVAMDENRVIGKDNRLPWHLPADLAYFKRVTMGHAIVMGRKTFEAIGRPLPGRDNVVVTRNRSFRPEGCLVLHSLEEVKQWIASRADEVFIIGGAELFRATMPIVDRLYVTKIFASFPGDTFYPPISDDEWEIVSYTPGGKDEKNPYEHAFIIYER\n",
      "MSA index 69 matches PDB\n",
      "\n",
      "\n",
      "#--------------------------------------------------------------#\n",
      "#---------- Begin Pre-Processing MSA --------------------------#\n",
      "#--------------------------------------------------------------#\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "#-------------------------Remove Gaps--------------------------#\n",
      "Shape of s is :  (7750, 159)\n",
      "in range indices:  [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17\n",
      "  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35\n",
      "  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53\n",
      "  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71\n",
      "  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89\n",
      "  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107\n",
      " 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125\n",
      " 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143\n",
      " 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158]\n",
      "s[tpdb] shape is  (159,)\n",
      "though s still has gaps, s[69] does not:\n",
      " ['M' 'I' 'S' 'H' 'I' 'V' 'A' 'M' 'D' 'E' 'N' 'R' 'V' 'I' 'G' 'K' 'D' 'N'\n",
      " 'R' 'L' 'P' 'W' 'H' 'L' 'P' 'A' 'D' 'L' 'A' 'Y' 'F' 'K' 'R' 'V' 'T' 'M'\n",
      " 'G' 'H' 'A' 'I' 'V' 'M' 'G' 'R' 'K' 'T' 'F' 'E' 'A' 'I' 'G' 'R' 'P' 'L'\n",
      " 'P' 'G' 'R' 'D' 'N' 'V' 'V' 'V' 'T' 'N' 'R' 'S' 'F' 'R' 'P' 'E' 'G' 'C'\n",
      " 'L' 'V' 'L' 'H' 'S' 'L' 'E' 'E' 'V' 'K' 'Q' 'W' 'I' 'A' 'S' 'R' 'A' 'D'\n",
      " 'E' 'V' 'F' 'I' 'I' 'G' 'G' 'A' 'E' 'L' 'F' 'R' 'A' 'T' 'M' 'P' 'I' 'V'\n",
      " 'D' 'R' 'L' 'Y' 'V' 'T' 'K' 'I' 'F' 'A' 'S' 'F' 'P' 'G' 'D' 'T' 'F' 'Y'\n",
      " 'P' 'P' 'I' 'S' 'D' 'D' 'E' 'W' 'E' 'I' 'V' 'S' 'Y' 'T' 'P' 'G' 'G' 'K'\n",
      " 'D' 'E' 'K' 'N' 'P' 'Y' 'E' 'H' 'A' 'F' 'I' 'I' 'Y' 'E' 'R']\n",
      "s shape is  (7750, 159)\n",
      "#--------------------------------------------------------------#\n",
      "\n",
      "\n",
      "removing non aligned (lower case) columns in subject sequence:\n",
      " [] \n",
      "\n",
      "#--------------------- Removing Duplicate Sequences -----------#\n",
      "...found 205 duplicates! (Removing...)\n",
      "#--------------------------------------------------------------#\n",
      "\n",
      "\n",
      "After removing bad sequences, tpdb is now  68\n",
      "1010\n",
      "After removing bad sequences, tpdb is now  54\n",
      "\n",
      "After removing bad sequences...\n",
      "tpdb (s_ipdb) is :  54\n",
      "(6535, 159)\n",
      "Found bad columns := [  0  88 143 144 157 158]\n",
      "Found conserved columns (80% repetition):\n",
      " [  6  13  14  21  26  30  34  41  42  43  45  48  53  54  56  60  95  96\n",
      " 113 122 126]\n",
      "We remove conserved and bad columns with, at the following indices (len 27):\n",
      " [  0   6  13  14 143 144  21  88  26 157  30 158  95  96  34  41  42  43\n",
      "  45  48 113  53  54  56 122  60 126]\n",
      "Removed Columns...\n",
      "Done Pre-Processing MSA\n",
      "#--------------------------------------------------------------#\n",
      "\n",
      "\n",
      "\n",
      "#--------------------------------------------------------------#\n",
      "Final s (MSA) has shape:  (6535, 132)\n",
      "s_index (length=132) = \n",
      " [  1   2   3   4   5   7   8   9  10  11  12  15  16  17  18  19  20  22\n",
      "  23  24  25  27  28  29  31  32  33  35  36  37  38  39  40  44  46  47\n",
      "  49  50  51  52  55  57  58  59  61  62  63  64  65  66  67  68  69  70\n",
      "  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  89\n",
      "  90  91  92  93  94  97  98  99 100 101 102 103 104 105 106 107 108 109\n",
      " 110 111 112 114 115 116 117 118 119 120 121 123 124 125 127 128 129 130\n",
      " 131 132 133 134 135 136 137 138 139 140 141 142 145 146 147 148 149 150\n",
      " 151 152 153 154 155 156]\n",
      "pdb_s_index (length=132) = \n",
      " [  1   2   3   4   5   7   8   9  10  11  12  15  16  17  18  19  20  22\n",
      "  23  24  25  27  28  29  31  32  33  35  36  37  38  39  40  44  46  47\n",
      "  49  50  51  52  55  57  58  59  61  62  64  65  66  67  68  69  70  71\n",
      "  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  90\n",
      "  91  92  93  94  95  98  99 100 101 102 103 104 105 106 107 108 109 110\n",
      " 111 112 113 115 116 117 118 119 120 121 122 124 125 126 128 129 130 131\n",
      " 132 133 134 135 136 137 138 139 140 141 142 143 146 147 148 149 150 151\n",
      " 152 153 154 155 156 157]\n",
      "Ref Seq (shape= (132,) ): \n",
      " ['I' 'S' 'H' 'I' 'V' 'M' 'D' 'E' 'N' 'R' 'V' 'K' 'D' 'N' 'R' 'L' 'P' 'H'\n",
      " 'L' 'P' 'A' 'L' 'A' 'Y' 'K' 'R' 'V' 'M' 'G' 'H' 'A' 'I' 'V' 'K' 'F' 'E'\n",
      " 'I' 'G' 'R' 'P' 'G' 'D' 'N' 'V' 'V' 'T' 'N' 'R' 'S' 'F' 'R' 'P' 'E' 'G'\n",
      " 'C' 'L' 'V' 'L' 'H' 'S' 'L' 'E' 'E' 'V' 'K' 'Q' 'W' 'I' 'A' 'S' 'R' 'D'\n",
      " 'E' 'V' 'F' 'I' 'I' 'A' 'E' 'L' 'F' 'R' 'A' 'T' 'M' 'P' 'I' 'V' 'D' 'R'\n",
      " 'L' 'Y' 'V' 'K' 'I' 'F' 'A' 'S' 'F' 'P' 'G' 'T' 'F' 'Y' 'P' 'I' 'S' 'D'\n",
      " 'D' 'E' 'W' 'E' 'I' 'V' 'S' 'Y' 'T' 'P' 'G' 'G' 'E' 'K' 'N' 'P' 'Y' 'E'\n",
      " 'H' 'A' 'F' 'I' 'I' 'Y']\n",
      "Finishing by converting amino acid letters to equvalent number form\n",
      "#--------------------------------------------------------------#\n",
      "\n",
      "\n",
      "\n",
      "Saving updated Prody Search DataFrame:\n",
      "    Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1  Unnamed: 0.1.1.1  \\\n",
      "0           0             0               0                 0   \n",
      "1           1             1               1                 1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1  Unnamed: 0.1.1.1.1.1  Unnamed: 0.1.1.1.1.1.1  \\\n",
      "0                   0                     0                       0   \n",
      "1                   1                     1                       1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1.1.1.1  Unnamed: 0.1.1.1.1.1.1.1.1  \\\n",
      "0                         0                           0   \n",
      "1                         1                           1   \n",
      "\n",
      "   Unnamed: 0.1.1.1.1.1.1.1.1.1  ...  ali_end ali_start bitscore  end  \\\n",
      "0                             0  ...      160         1   215.11  160   \n",
      "1                             1  ...      160         1   215.08  160   \n",
      "\n",
      "    cond_evalue    ind_evalue    evidence hmm_end hmm_start start  \n",
      "0  2.300000e-68  4.500000e-64  hmmer v3.0     160         1     1  \n",
      "1  2.300000e-68  4.600000e-64  hmmer v3.0     160         1     1  \n",
      "\n",
      "[2 rows x 30 columns]\n",
      "Done Preprocessing Data.....\n",
      "Number of residue positions: 132\n",
      "Number of sequences: 6535\n",
      "N_effective  4815.092888344703\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n\\n\\nLooping through Prody Search DataFrame:\\n\", prody_df.head())\n",
    "rows_to_drop = []\n",
    "for ir, pdb2msa_row in enumerate(prody_df.iterrows()):\n",
    "    print('\\n\\nGetting msa with following pdb2msa entry:\\n', pdb2msa_row)\n",
    "    #try:\n",
    "    dp_result =  data_processing(data_path, prody_df.iloc[pdb2msa_row[0]], gap_seqs=0.2, gap_cols=0.2, prob_low=0.004,\n",
    "                               conserved_cols=0.8, printing=True, out_dir=processed_data_dir, pdb_dir=pdb_dir, letter_format=False,\n",
    "                               remove_cols=True, create_new=True, n_cpu=min(2, n_cpus))\n",
    "    if dp_result is not None:\n",
    "        [s0, removed_cols, s_index, tpdb, pdb_s_index] = dp_result\n",
    "        break\n",
    "    else:\n",
    "        rows_to_drop.append(ir)\n",
    "        continue\n",
    "\n",
    "\n",
    "\n",
    "pdb_id = pdb2msa_row[1]['PDB ID']\n",
    "pfam_id = pdb2msa_row[1]['Pfam']\n",
    "chain = pdb2msa_row[1]['Chain']\n",
    "# update Prody search DF (use same filename as pdb2msa() in data_processing\n",
    "prody_df = prody_df.drop(rows_to_drop)\n",
    "print(\"\\n\\n\\nSaving updated Prody Search DataFrame:\\n\", prody_df.head())\n",
    "prody_df.to_csv('%s/%s_pdb_df.csv' % (pdb_dir, pdb_id))\n",
    "\n",
    "if dp_result is None:\n",
    "    print('None of the available prody pdb search found matching alignments... Exiting..')\n",
    "    sys.exit()\n",
    "print('Done Preprocessing Data.....')\n",
    "\n",
    "\n",
    "\n",
    "# number of positions\n",
    "n_var = s0.shape[1]\n",
    "n_seq = s0.shape[0]\n",
    "\n",
    "print(\"Number of residue positions:\",n_var)\n",
    "print(\"Number of sequences:\",n_seq)\n",
    "\n",
    "# Compute effective number of sequences\n",
    "dst = distance.squareform(distance.pdist(s0, 'hamming'))\n",
    "theta = .2 # minimum necessary distance (theta = 1. - seq_identity_thresh)\n",
    "seq_ints = (dst < theta).sum(axis=1).astype(float)\n",
    "ma_inv = 1/((dst < theta).sum(axis=1).astype(float))  \n",
    "meff = ma_inv.sum()\n",
    "\n",
    "print('N_effective ', meff)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4955b6a0",
   "metadata": {},
   "source": [
    "## Use Expectation Reflection to calculate Mutual Information\n",
    "* Re-Define MSA using OneHot transformation to get binary variables for ER\n",
    "* Calculate in parallel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38372222",
   "metadata": {},
   "source": [
    "### OneHot tranformation and initialization of weight matix and fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "667932b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different amino acids at each position:\n",
      " [ 6 16 12  7  8 15  8 17 12 14  9 19 11 10 13  5  5 14  9  5 11  9 12 10\n",
      "  9 10 16 17 11 12 13  5  4  9  7 11  5 10 12  9  9 17  8  7  5  2 11 18\n",
      " 18 20 20 20 14 13 15 16 15 14 14  7 12 15 12  8 15 15 21 19 16 16 20 15\n",
      " 17 13 11  8 10 11 13  4  3 13 15 12 11 13 19  9 11 17  7 14  8 16  3 15\n",
      " 19 15 14 15  7  7 15  7 17 10 16 20 17 18  5 17 16 17 19 19 18 21 21 19\n",
      " 15 17 13 16 13 14 10 14  9 18 14  5]\n",
      "\n",
      "Total number of variables: 1664\n"
     ]
    }
   ],
   "source": [
    "# --- Get number of binary aa-pair-position variables ie aa-aa pair at bosition i (yes/no) ---- #\n",
    "# number of aminoacids at each position #\n",
    "mx = np.array([len(np.unique(s0[:,i])) for i in range(n_var)]) #\n",
    "print(\"Number of different amino acids at each position:\\n\",mx) #\n",
    "\n",
    "mx_cumsum = np.insert(mx.cumsum(),0,0) \n",
    "i1i2 = np.stack([mx_cumsum[:-1],mx_cumsum[1:]]).T  \n",
    "mx_sum = mx.sum() \n",
    "print(\"\\nTotal number of variables:\",mx_sum) \n",
    " # number of bias term \n",
    "n_linear = mx_sum - n_var \n",
    "onehot_encoder = OneHotEncoder(sparse=False,categories='auto') \n",
    "# s is OneHot encoder format, s0 is original sequnce matrix #\n",
    "s = onehot_encoder.fit_transform(s0)\n",
    "# --------------------------------------------------------------------------------------------- #\n",
    "\n",
    "\n",
    "# Define wight matrix with variable for each possible amino acid at each sequence position\n",
    "w = np.zeros((mx.sum(),mx.sum())) \n",
    "h0 = np.zeros(mx.sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33a077f3",
   "metadata": {},
   "source": [
    "### Define Expectation Reflection Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0e7d27b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Expectation Reflection\n",
    "#=========================================================================================\n",
    "def predict_w(s,i0,i1i2,niter_max,l2):\n",
    "    #print('i0:',i0)\n",
    "    i1,i2 = i1i2[i0,0],i1i2[i0,1]\n",
    "\n",
    "    x = np.hstack([s[:,:i1],s[:,i2:]])\n",
    "    y = s[:,i1:i2]\n",
    "\n",
    "    h01,w1 = ER.fit(x,y,niter_max,l2)\n",
    "\n",
    "    return h01,w1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e30590d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/cresswellclayec/conda/envs/DCA_ER/lib/python3.9/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  warnings.warn(\n",
      "/data/cresswellclayec/conda/envs/DCA_ER/lib/python3.9/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  warnings.warn(\n",
      "/data/cresswellclayec/conda/envs/DCA_ER/lib/python3.9/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  warnings.warn(\n",
      "/data/cresswellclayec/conda/envs/DCA_ER/lib/python3.9/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  warnings.warn(\n",
      "/data/cresswellclayec/conda/envs/DCA_ER/lib/python3.9/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  warnings.warn(\n",
      "/data/cresswellclayec/conda/envs/DCA_ER/lib/python3.9/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  warnings.warn(\n",
      "/data/cresswellclayec/conda/envs/DCA_ER/lib/python3.9/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  warnings.warn(\n",
      "/data/cresswellclayec/conda/envs/DCA_ER/lib/python3.9/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "run time: 199.44644566701027\n",
      "Dimensions of w:  (1664, 1664)\n",
      "[[0.         0.1192203  0.01825799 ... 0.00586072 0.00540024 0.00560581]\n",
      " [0.1192203  0.         0.02614052 ... 0.00526424 0.00968795 0.01024421]\n",
      " [0.01825799 0.02614052 0.         ... 0.00559539 0.00477334 0.00758165]\n",
      " ...\n",
      " [0.00586072 0.00526424 0.00559539 ... 0.         0.03711063 0.10005056]\n",
      " [0.00540024 0.00968795 0.00477334 ... 0.03711063 0.         0.09601378]\n",
      " [0.00560581 0.01024421 0.00758165 ... 0.10005056 0.09601378 0.        ]]\n",
      "(132, 132)\n"
     ]
    }
   ],
   "source": [
    "import expectation_reflection as ER\n",
    "# parallel\n",
    "start_time = timeit.default_timer()\n",
    "#res = Parallel(n_jobs = 4)(delayed(predict_w)\\\n",
    "#res = Parallel(n_jobs = 8)(delayed(predict_w)\\\n",
    "res = Parallel(n_jobs = n_cpus-2)(delayed(predict_w)\\\n",
    "        (s,i0,i1i2,niter_max=10,l2=100.0)\\\n",
    "        for i0 in range(n_var))\n",
    "\n",
    "run_time = timeit.default_timer() - start_time\n",
    "print('run time:',run_time)\n",
    "for i0 in range(n_var):\n",
    "    i1,i2 = i1i2[i0,0],i1i2[i0,1]\n",
    "\n",
    "    h01 = res[i0][0]\n",
    "    w1 = res[i0][1]\n",
    "\n",
    "    h0[i1:i2] = h01\n",
    "    w[:i1,i1:i2] = w1[:i1,:]\n",
    "    w[i2:,i1:i2] = w1[i1:,:]\n",
    "\n",
    "# make w symmetric\n",
    "w = (w + w.T)/2.\n",
    "\n",
    "# Verify that w is symmetric (sanity test)\n",
    "print(\"Dimensions of w: \",w.shape)\n",
    "\n",
    "di = direct_info(s0,w)\n",
    "print(di)\n",
    "print(di.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e9e90f49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12, 121) 0.29401320305773226\n",
      "(49, 50) 0.1916623361857236\n",
      "(35, 55) 0.16986300336871032\n",
      "(57, 72) 0.16159017901821776\n",
      "(39, 58) 0.15106770779132442\n"
     ]
    }
   ],
   "source": [
    "# Verify that w is symmetric (sanity test)\n",
    "\n",
    "\n",
    "np.save(\"%s/%s_%s_ER_di.npy\" % (out_dir, pdb_id, pfam_id), di)\n",
    "\n",
    "# Convert DI matrix to DI score dictionary\n",
    "DI_ER_pydca = tools.scores_matrix2dict(di, s_index)\n",
    "for site_pair, score in DI_ER_pydca[:5]:\n",
    "    print(site_pair, score)\n",
    "import pickle as pkl\n",
    "with open(\"%s/%s_%s_ER_di_pydca.npy\" % (out_dir, pdb_id, pfam_id), \"wb\") as f_pydca:\n",
    "    pkl.dump(DI_ER_pydca,f_pydca )\n",
    "f_pydca.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "457629ee",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
